### PR TITLE
[SLP] Fix PoisonValue in the argument VL of setOperand.

### PR DESCRIPTION
**Pull request by HanKuanChen**
In addition, Scalars is already in the struct, we don't need to pass VL here.

reference: https://github.com/llvm/llvm-project/pull/113880

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -3338,14 +3338,13 @@ class BoUpSLP {
       copy(OpVL, Operands[OpIdx].begin());
     }
 
-    /// Set this bundle's operand from \p VL.
-    void setOperand(ArrayRef<Value *> VL, const BoUpSLP &R,
-                    bool RequireReorder = false) {
-      VLOperands Ops(VL, R);
+    /// Set this bundle's operand from Scalars.
+    void setOperand(const BoUpSLP &R, bool RequireReorder = false) {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3342]: Better to pass first instruction as an argument rather than redo the search for it

**Comment by alexey-bataev** [comment]
Your original patch was reverted, better to revive the previous review and incorporate this fix into the previous review

**Comment by HanKuanChen** [comment]
> Your original patch was reverted, better to revive the previous review and incorporate this fix into the previous review

I will cherry pick this patch once this patch is approved.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -3338,14 +3341,13 @@ class BoUpSLP {
...
+      auto *I0 = cast<Instruction>(*find_if(Scalars, IsaPred<Instruction>));
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3346]: You still looking for the instruction here, better to pass it as an argument or make it part of VLOperands class, just like Scalars

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3346]: `I0` is passed to `VLOperands Ops(Scalars, R, I0);` to reduce the search. Is this not good?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3346]: You still redo search here

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -2542,13 +2541,11 @@ class BoUpSLP {
 
   public:
     /// Initialize with all the operands of the instruction vector \p RootVL.
-    VLOperands(ArrayRef<Value *> RootVL, const BoUpSLP &R)
+    VLOperands(ArrayRef<Value *> RootVL, Instruction *VL0, const BoUpSLP &R)
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:2544]: Make VL0 a member of the сlass, it is used in too many cases

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:2544]: VL0 is only used by the constructor of VLOperands. I think you mean "Make VL0 a member of TreeEntry"?

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:2544]: If we need to put VL0 into TreeEntry, I suggest we open the original PR (https://github.com/llvm/llvm-project/pull/113880#issuecomment-2523263864) and review there.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 4]
```diff
@@ -3338,14 +3335,13 @@ class BoUpSLP {
       copy(OpVL, Operands[OpIdx].begin());
     }
 
-    /// Set this bundle's operand from \p VL.
-    void setOperand(ArrayRef<Value *> VL, const BoUpSLP &R,
+    /// Set this bundle's operand from Scalars.
+    void setOperand(Instruction *VL0, const BoUpSLP &R,
                     bool RequireReorder = false) {
-      VLOperands Ops(VL, R);
+      VLOperands Ops(Scalars, VL0, R);
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3341]: Instead of passing `VL0` here, I think you can safely use MainOp member instruction

**Review by alexey-bataev — APPROVED** [review]

**Pull request merged**